### ISSUE TITLE
feature request: back-button outside of ion-nav

### ISSUE THREAD

**Issue by brendensoares** [ionic-team/ionic-framework#22941]
# Feature Request
Ionic version:
5.x

# Describe the Feature Request
I want to use `ion-nav` with a common header bar (with back button and title).

I have a nearly working solution, but `ion-back-button` only works as a descendant of an `ion-nav`. It seems it was designed for each pushed nav item to have its own header and back button. This is not ideal because the transition animation is not fluid during the nav push, it increases code duplication by having `ion-header`s/tool bars in each nav stack item/screen, and it makes it difficult to have a common/shared UI.

# Describe Preferred Solution
Don't require `ion-back-button` to be within an `ion-nav`. Perhaps add an optional property to target a specific `ion-nav`.

# Describe Alternatives
The only other option right now is to duplicate the `ion-header`/tool bar in each nav stack item/screen.

# Related Code
https://github.com/ionic-team/ionic-framework/blob/753fd2f910e718438e3ac918e2f129501d8e9791/core/src/components/back-button/back-button.tsx#L108

# Additional Context
A side note, the docs for `ion-nav` and the related components seems lacking and possibly out of date.

I'm reopening #22454 to reply to @liamdebeasi. I am looking to reproduce iOS's native [`UINavController`](https://developer.apple.com/documentation/uikit/uinavigationcontroller) which is responsible for the nav stack of screens/views/pages (aka `ViewController`s) starting with a root screen/view/page. Each child screen does not manage the title bar directly. It delegates its needs by defining the title text, left and right button.

In Ionic, I have to provide the title bar and back button for each screen in the nav stack. This is less than ideal as it is redundant boilerplate code.

I've not toyed with this issue for some months, but I am now returning to the code to see if I can start a new app code base. So hopefully my comments are still relevant. I've tried to adjust my original issue details above to be more clear. I'll try to pay attention to replies to this issue to avoid getting it closed again. Thanks.

**Comment by brendensoares** [comment]
Here's my exact code attempt at a "nav controller" component:
```typescript
import { Component, Host, h, State } from '@stencil/core';

interface FlowSection {
  name : string;
  steps: FlowStep[];
}

interface FlowStep {
  name         : string;
  componentName: string;
}

function newSection(name: string, steps: FlowStep[]): FlowSection {
  return { name, steps };
}

function newStep(name: string, componentName: string): FlowStep {
  return { name, componentName };
}

@Component({
  tag: 'page-onboard',
  styleUrl: 'page-onboard.css',
})
export class PageOnboard {

  navEl: any; // TOOD: consider an explicit interface type

  sections: FlowSection[] = [
    newSection("Company", [
      newStep("Company Profile", "page-onboard-company")
    ]),
    newSection("Services", [
      newStep("Activate Services", "page-onboard-services"),
      newStep("Availability", "page-onboard-service-availability"),
      newStep("Capacity", "page-onboard-service-capacity"),
      newStep("Capacity Overrides", "page-onboard-service-capacity-overrides"),
      newStep("Pricing", "page-onboard-service-pricing"),
      newStep("Addons", "page-onboard-service-pricing-addons"),
      newStep("Discounts", "page-onboard-service-pricing-discounts"),
    ]),
    newSection("Storefront", [
      newStep("Storefront Setup", "page-onboard-storefront")
    ])
  ]

  exitPage: string = "page-onboard-complete";

  @State() currentSectionIndex: number = 0;
  @State() currentStepIndex: number = 0;

  nextStep() {
    let section = this.sections[this.currentSectionIndex];

    if (section.steps.length-1 === this.currentStepIndex) {
      // Last step passed, point to next section
      this.currentSectionIndex++;
      if (this.sections.length-1 === this.currentSectionIndex-1) {
        // Last section passed, navto exit page
        this.navEl.setRoot(this.exitPage);
        return;
      }
      this.currentStepIndex = 0;
      section = this.sections[this.currentSectionIndex];
    } else {
      // More steps to come
      this.currentStepIndex++;
    }
    let step = section.steps[this.currentStepIndex];

    // Nav to next step
    this.navEl.push(step.componentName, {nextStep: () => this.nextStep()});
  }

  componentDidLoad() {
    this.navEl = document.querySelector('ion-nav#onboard-nav');
  }

  render() {
    return (
      <Host no-router>
        <ion-header>
          <ion-toolbar color="primary">
            <ion-buttons slot="start">
            </ion-buttons>
            <ion-title>{this.sections.length === this.currentSectionIndex ? 'Welcome Aboard!' : 'Onboarding'}</ion-title>
          </ion-toolbar>
          <ion-row>
            { this.sections.map( (sec, i) => <ion-col>
              <ion-progress-bar color="secondary" value={this.currentSectionIndex > i ? 1 : (this.currentSectionIndex == i ? this.currentStepIndex / (sec.steps.length) : 0)} />
            </ion-col>
            )}
          </ion-row>
        </ion-header>

        <ion-content>
          <ion-nav id="onboard-nav" root={this.sections[0].steps[0].componentName} rootParams={{nextStep: () => this.nextStep()}}/>
        </ion-content>
      </Host>
    );
  }

}
```

and here is one of the "steps" aka nav stack items:
```typescript
import { Component, Host, h, Prop } from '@stencil/core';

@Component({
  tag: 'page-onboard-company',
  styleUrl: 'page-onboard-company.css',
})
export class PageOnboardCompany {

  @Prop() nextStep: Function;

  handleSubmit(e) {
    e.preventDefault();
    this.nextStep();
  }

  render() {
    return (
      <Host>
        <ion-card>
          <form onSubmit={(e) => this.handleSubmit(e)}>
            <ion-item lines="none">
              <p>company</p>
            </ion-item>
            <ion-item>
              <ion-label position="floating">Password</ion-label>
              <ion-input type="password" minlength={8} pattern="[A-Za-z0-9]+"></ion-input>
            </ion-item>
            <rf-button />
          </form>
        </ion-card>
      </Host>
    );
  }

}
```

Hopefully this makes it more obvious what I am ask for and what I am trying to accomplish. It's a multi-step workflow that the user goes through. I am trying to centralize the workflow management in this "nav controller" component so that the child "steps" don't need to be concerned with what step they are in the flow (eg which step is next, etc).

**Comment by brendensoares** [comment]
NOTE: I'd like to have a `ion-back-button` in the `PageOnboard`'s `ion-header` and NOT in any of the child step components such as `PageOnboardCompany`.

**Comment by brendensoares** [comment]
Here's a [video example](https://youtu.be/ikFVfcJY6VE?t=246)

**Comment by liamdebeasi** [comment]
Thanks for the issue. Can you provide a minimal GitHub repo that shows what you are trying to accomplish?

In previous versions of Ionic Framework, we did not require developers to add an `ion-back-button` to each view. While it reduced the boilerplate code, it caused several other issues that made the development experience frustrating. Due to these previous challenges I do not think we are likely to add this feature, but I am interested in seeing if we can find another way to address your use case.

**Comment by brendensoares** [comment]
It seems to me that all that needs to be done is that the `ion-back-button` does not require it be a descendant of an `ion-nav` instance so that the developer can place the side by side as they see fit.

**Comment by liamdebeasi** [comment]
The main concern the team has is that if we were to do this we would need to extend that capability to our routing integrations as well, which means this needs to work with Vue Router, React Router, and Angular Router. If `ion-back-button` works outside of `ion-nav` I think it would be reasonable to assume it would work outside of other navigation contexts as well. While it certainly is not impossible, it would require a significant amount of work that the team is not comfortable taking on. We also feel that this feature could re-introduce many of the back button issues developers experienced in previous versions of Ionic Framework.

Like I said before, I am interested in exploring alternatives that may fit your use case. Specifically, I am interested in the following line that you noted:

> This is not ideal because the transition animation is not fluid during the nav push, it increases code duplication by having ion-headers/tool bars in each nav stack item/screen, and it makes it difficult to have a common/shared UI.

If you could provide a GitHub repo that shows the problems you are running into, this would be really helpful. Thanks!

**Comment by brendensoares** [comment]
I provided relevant code files to fully demonstrate the issue. Perhaps I can work up a minimal example repo soon, but I don't see how removing the parent `ion-nav` requirement would affect any other use cases. Can you clarify the issues there?

Also, can you link to the prior back button issues so I can expand my perspective a bit more?

**Comment by liamdebeasi** [comment]
There are two main issues here. The first is that decoupling `ion-back-button` from its navigation context is not as straightforward as it seems. Most of the assumptions the framework makes when handling view management, transitions, etc is based on the fact that `ion-back-button` exists inside of its navigation context. Changing this assumption involves a significant amount of work just inside the framework which is not something we are looking to do at the moment.

As an example, the iOS transition assumes that any `ion-back-button` usage will be in the entering and leaving views directly: https://github.com/ionic-team/ionic-framework/blob/master/core/src/utils/transition/ios.transition.ts#L324-L331 Changing where `ion-back-button` is used would influence things as low level as how the transitions work.

The second issue is that if we change this assumption, we also need to consider how this assumption influences things such as our routing integrations. It would be one thing if we only had to worry about our Vanilla JS router, but we also have to consider how this would influence our Vue, Angular, and React integrations.

> Also, can you link to the prior back button issues so I can expand my perspective a bit more?

Honestly I am not sure where those GitHub issues are as it was before I joined the company. I used Ionic Framework v3 at previous companies I worked at and ran into issues such as the back button being added to views when I did not want it to as well as the back button not showing up when it should be. What I do recall is that the requirement that `ion-back-button` be added to each view was directly influenced by these challenges.

> Perhaps I can work up a minimal example repo soon

That would be great, thank you! 😊
